In [2]:
import pandas as pd

In [3]:
# read csv filt in to a dataframe which is lending club data between 2007 and 2011
loans_2011 = pd.read_csv('LoanStats3e.csv', encoding = "ISO-8859-1")

D:\Users\KDH\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (0,47) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
# display first row of the dataframe
loans_2011.head(1)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     5000.0       5000.0           4975.0   36 months   

   int_rate  installment grade sub_grade       ...        recoveries  \
0    0.1065       162.87     B        B2       ...               0.0   

  collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                     0.0     Jan-2015           171.62          NaN   

  last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog  \
0           Aug-2018                        0.0                         NaN   

   policy_code application_type  
0          1.0       Individual  

[1 rows x 53 columns]

In [5]:
# display the shape of the dataframe
loans_2011.shape

(42538, 53)

** it has 53 columns and 42538 rows **

** display the columns of the dataframe ** 
- reference: LCDataDictionary.xlsx 

In [6]:
loans_2011.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title',
       'emp_length', 'home_ownership', 'annual_inc', 'verification_status',
       'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose',
       'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs',
       'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp',
       'total_rec_int', 'total_rec_late_fee', 'recoveries',
       'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt',
       'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med',
       'mths_since_last_major_derog', 'policy_code', 'application_type'],
      dtype='object')

- id	object		A unique LC assigned ID for the loan listing.
- member_id	float64		A unique LC assigned Id for the borrower member.
- loan_amnt	float64		The listed amount of the loan applied for by the borrower.
- funded_amnt	float64		The total amount committed to that loan at that point in time.
- funded_amnt_inv	float64		The total amount committed by investors for that loan at that point in time.
- term	object		The number of payments on the loan. Values are in months and can be either 36 or 60.
- int_rate	object		Interest Rate on the loan
- installment	float64		The monthly payment owed by the borrower if the loan originates.
- grade	object		LC assigned loan grade
- sub_grade	object		LC assigned loan subgrade
- emp_title	object		The job title supplied by the Borrower when applying for the loan.
- emp_length	object		Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.
- home_ownership	object		The home ownership status provided by the borrower during registration. Our values are: RENT, OWN, MORTGAGE, OTHER.
- annual_inc	float64		The self-reported annual income provided by the borrower during registration.
- verification_status	object	Verified	Indicates if income was verified by LC, not verified, or if the income source was verified
- issue_d	object		The month which the loan was funded
- loan_status	object		Current status of the loan
- pymnt_plan	object		Indicates if a payment plan has been put in place for the loan
- purpose	object		A category provided by the borrower for the loan request.
- url:URL for the LC page with listing data.
- desc:Loan description provided by the borrower
- mths_since_last_delinq : The number of months since the borrower's last delinquency.
- mths_since_last_record : The number of months since the last public record.
- mths_since_last_major_dero g: Months since most recent 90-day or worse rating at time of application for the secondary applicant
- next_pymnt_d :Next scheduled payment date

+ After analyzing each column, we can conclude that the following features need to be removed:


- id: randomly generated field by Lending Club for unique identification purposes only
- member_id: also a randomly generated field by Lending Club for unique identification purposes only
- funded_amnt: leaks data from the future (after the loan is already started to be funded)
- funded_amnt_inv: also leaks data from the future (after the loan is already started to be funded)
- grade: contains redundant information as the interest rate column (int_rate)
- sub_grade: also contains redundant information as the interest rate column (int_rate)
- emp_title: requires other data and a lot of processing to potentially be useful
- issue_d: leaks data from the future (after the loan is already completed funded)
- url: empty
- desc: desc is not much imformative for machine learning
- next_pymnt_d: payment date is not much relevant

In [7]:
loans_2011 = loans_2011.drop(["id", "member_id", "funded_amnt", "funded_amnt_inv", "grade", "sub_grade", "emp_title", "issue_d","url","desc","next_pymnt_d","mths_since_last_major_derog"], axis=1)

- name	dtype	first value	description
- title	object	Computer	The loan title provided by the borrower
- zip_code	object	860xx	The first 3 numbers of the zip code provided by the borrower in the loan application.
- addr_state	object	AZ	The state provided by the borrower in the loan application
- dti	float64	27.65	A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.
- delinq_2yrs	float64	0	The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years
- earliest_cr_line	object	Jan-1985	The month the borrower's earliest reported credit line was opened
- inq_last_6mths	float64	1	The number of inquiries in past 6 months (excluding auto and mortgage inquiries)
- open_acc	float64	3	The number of open credit lines in the borrower's credit file.
- pub_rec	float64	0	Number of derogatory public records
- revol_bal	float64	13648	Total credit revolving balance
- revol_util	object	83.7%	Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit.
- total_acc	float64	9	The total number of credit lines currently in the borrower's credit file
- initial_list_status	object	f	The initial listing status of the loan. Possible values are – W, F
- out_prncp	float64	0	Remaining outstanding principal for total amount funded
- out_prncp_inv	float64	0	Remaining outstanding principal for portion of total amount funded by investors
- total_pymnt	float64	5863.16	Payments received to date for total amount funded
- total_pymnt_inv	float64	5833.84	Payments received to date for portion of total amount funded by investors
- total_rec_prncp	float64	5000	Principal received to date

- Within this group of columns, we need to drop the following columns:


- zip_code: redundant with the addr_state column since only the first 3 digits of the 5 digit zip code are visible (which only can be used to identify the state the borrower lives in)
- out_prncp: leaks data from the future, (after the loan already started to be paid off)
- out_prncp_inv: also leaks data from the future, (after the loan already started to be paid off)
- total_pymnt: also leaks data from the future, (after the loan already started to be paid off)
- total_pymnt_inv: also leaks data from the future, (after the loan already started to be paid off)
- total_rec_prncp: also leaks data from the future, (after the loan already started to be paid off)

In [8]:
loans_2011 = loans_2011.drop(["zip_code", "out_prncp", "out_prncp_inv", "total_pymnt", "total_pymnt_inv", "total_rec_prncp"], axis=1)

- name	dtype	first value	description

- total_rec_int	float64	863.16	Interest received to date
- total_rec_late_fee	float64	0	Late fees received to date
- recoveries	float64	0	post charge off gross recovery
- collection_recovery_fee	float64	0	post charge off collection fee
- last_pymnt_d	object	Jan-2015	Last month payment was received
- last_pymnt_amnt	float64	171.62	Last total payment amount received
- last_credit_pull_d	object	Jun-2016	The most recent month LC pulled credit for this loan
- collections_12_mths_ex_med	float64	0	Number of collections in 12 months excluding medical collections
- policy_code	float64	1	publicly available policy_code=1 new products not publicly available policy_code=2
- application_type	object	INDIVIDUAL	Indicates whether the loan is an individual application or a joint application with two co-borrowers
- acc_now_delinq	float64	0	The number of accounts on which the borrower is now delinquent.
- chargeoff_within_12_mths	float64	0	Number of charge-offs within 12 months
- delinq_amnt	float64	0	The past-due amount owed for the accounts on which the borrower is now delinquent.
- pub_rec_bankruptcies	float64	0	Number of public record bankruptcies
- tax_liens	float64	0	Number of tax liens


- In the last group of columns, we need to drop the following columns:


- total_rec_int: leaks data from the future, (after the loan already started to be paid off),
- total_rec_late_fee: also leaks data from the future, (after the loan already started to be paid off),
- recoveries: also leaks data from the future, (after the loan already started to be paid off),
- collection_recovery_fee: also leaks data from the future, (after the loan already started to be paid off),
- last_pymnt_d: also leaks data from the future, (after the loan already started to be paid off),
- last_pymnt_amnt: also leaks data from the future, (after the loan already started to be paid off).

In [9]:
loans_2011 = loans_2011.drop(["total_rec_int", "total_rec_late_fee", "recoveries", "collection_recovery_fee", "last_pymnt_d", "last_pymnt_amnt"], axis=1)

In [10]:
print(loans_2011.head(3))
print(loans_2011.shape)

   loan_amnt        term  int_rate  installment emp_length home_ownership  \
0     5000.0   36 months    0.1065       162.87  10+ years           RENT   
1     2500.0   60 months    0.1527        59.83   < 1 year           RENT   
2     2400.0   36 months    0.1596        84.33  10+ years           RENT   

   annual_inc verification_status  loan_status pymnt_plan        ...         \
0     24000.0            Verified   Fully Paid          n        ...          
1     30000.0     Source Verified  Charged Off          n        ...          
2     12252.0        Not Verified   Fully Paid          n        ...          

  open_acc pub_rec revol_bal  revol_util  total_acc initial_list_status  \
0      3.0     0.0   13648.0       0.837        9.0                   f   
1      3.0     0.0    1687.0       0.094        4.0                   f   
2      2.0     0.0    2956.0       0.985       10.0                   f   

   last_credit_pull_d  collections_12_mths_ex_med  policy_code  \
0      

In [11]:
print(loans_2011['loan_status'].value_counts())

Fully Paid                                             34116
Charged Off                                             5670
Does not meet the credit policy. Status:Fully Paid      1988
Does not meet the credit policy. Status:Charged Off      761
Name: loan_status, dtype: int64


### Loan Status Meaning
- Fully Paid	34116	Loan has been fully paid off.
- Charged Off	5670	Loan for which there is no longer a reasonable expectation of further payments.
- Does not meet the credit policy. Status:Fully Paid	1988	While the loan was paid off, the loan application today would no longer meet the credit policy and wouldn't be approved on to the marketplace.
- Does not meet the credit policy. Status:Charged Off	761	While the loan was charged off, the loan application today would no longer meet the credit policy and wouldn't be approved on to the marketplace.


- Reference: https://help.lendingclub.com/hc/en-us

- In this case, I am, as an invester's perspective, interested in trying to predict which loans will be paid off on time and which ones won't be. Only the Fully Paid and Charged Off values describe the final outcome of the loan. The other values describe loans that are still on going and where the jury is still out on if the borrower will pay back the loan on time or not. While the Default status resembles the Charged Off status, in Lending Club's eyes, loans that are charged off have essentially no chance of being repaid while default ones have a small chance.


In [12]:
# replace categorical column with numerical values
loans_2011 = loans_2011[(loans_2011['loan_status'] == "Fully Paid") | (loans_2011['loan_status'] == "Charged Off")]

status_replace = {
    "loan_status" : {
        "Fully Paid": 1,
        "Charged Off": 0,
    }
}


In [13]:
# remove all rows from loans_2011 that contain values other than Fully Paid or Charged Off for the loan_status column.
loans_2011 = loans_2011.replace(status_replace)

In [14]:
# drop the columns that has only one value
orig_columns = loans_2011.columns

drop_columns = []
for col in orig_columns:
    col_series = loans_2011[col].dropna().unique()
    if len(col_series) == 1:
        drop_columns.append(col)
loans_2011 = loans_2011.drop(drop_columns, axis=1)
print(drop_columns)

['pymnt_plan', 'initial_list_status', 'collections_12_mths_ex_med', 'policy_code', 'application_type']


In [15]:
loans_2011.head()

loan_amnt        term  int_rate  installment emp_length home_ownership  \
0     5000.0   36 months    0.1065       162.87  10+ years           RENT   
1     2500.0   60 months    0.1527        59.83   < 1 year           RENT   
2     2400.0   36 months    0.1596        84.33  10+ years           RENT   
3    10000.0   36 months    0.1349       339.31  10+ years           RENT   
4     3000.0   60 months    0.1269        67.79     1 year           RENT   

   annual_inc verification_status  loan_status         purpose  \
0     24000.0            Verified            1     credit_card   
1     30000.0     Source Verified            0             car   
2     12252.0        Not Verified            1  small_business   
3     49200.0     Source Verified            1           other   
4     80000.0     Source Verified            1           other   

          ...         earliest_cr_line inq_last_6mths  mths_since_last_delinq  \
0         ...                 Jan-1985            1.0                     NaN   
1         ...                 Apr-1999            5.0                     NaN   
2         ...                 Nov-2001            2.0                     NaN   
3         ...                 Feb-1996            1.0                    35.0   
4         ...                 Jan-1996            0.0                    38.0   

   mths_since_last_record open_acc  pub_rec  revol_bal  revol_util  total_acc  \
0                     NaN      3.0      0.0    13648.0       0.837        9.0   
1                     NaN      3.0      0.0     1687.0       0.094        4.0   
2                     NaN      2.0      0.0     2956.0       0.985       10.0   
3                     NaN     10.0      0.0     5598.0       0.210       37.0   
4                     NaN     15.0      0.0    27783.0       0.539       38.0   

   last_credit_pull_d  
0            Aug-2018  
1            Oct-2016  
2            Jun-2017  
3            Apr-2016  
4            Apr-2018  

[5 rows x 24 columns]

** For Machine learning model, we need to remove all null values and values should be numerical. **

In [16]:
# it shows which columns have null values
null_counts = loans_2011.isnull().sum()
print(null_counts)

loan_amnt                     0
term                          0
int_rate                      0
installment                   0
emp_length                 1078
home_ownership                0
annual_inc                    0
verification_status           0
loan_status                   0
purpose                       0
title                        11
addr_state                    0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
mths_since_last_delinq    25727
mths_since_last_record    36995
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                   50
total_acc                     0
last_credit_pull_d            2
dtype: int64


In [20]:
# It shows that mths_since_last_delinq,mths_since_last_record,mths_since_last_major_derog columns contain 25727,36995,39786 values and title,revol_util,last_credit_pull_d columns each has 11,50,2 null values
# so remove columns entirely where more than 1% of the rows for that column contain a null value. 
# I decide not to drop emp_length column since it is much imformative
# In addition, we'll remove the remaining rows containing null values.

loans=loans_2011.drop(["mths_since_last_delinq","mths_since_last_record"],axis=1)
loans=loans.dropna(axis=0)
print(loans.dtypes.value_counts())

float64    12
object      9
int64       1
dtype: int64


In [21]:
# check if null values are removed properly
null_counts = loans.isnull().sum()
print(null_counts)

loan_amnt              0
term                   0
int_rate               0
installment            0
emp_length             0
home_ownership         0
annual_inc             0
verification_status    0
loan_status            0
purpose                0
title                  0
addr_state             0
dti                    0
delinq_2yrs            0
earliest_cr_line       0
inq_last_6mths         0
open_acc               0
pub_rec                0
revol_bal              0
revol_util             0
total_acc              0
last_credit_pull_d     0
dtype: int64


In [22]:
# treat emp_length column, firstly check the values in emp_length column
loans.emp_length.unique()

array(['10+ years', '< 1 year', '1 year', '3 years', '8 years', '9 years',
       '4 years', '5 years', '6 years', '2 years', '7 years'],
      dtype=object)

In [23]:
# Make dictionary and use it to replace emp_length with numerical values

mapping_dict = {
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    }
}
loans = loans.replace(mapping_dict)

In [24]:
# check the dtype of columns
object_columns_df = loans.select_dtypes(include=['object'])
print(object_columns_df.iloc[0])

term                     36 months
home_ownership                RENT
verification_status       Verified
purpose                credit_card
title                     Computer
addr_state                      AZ
earliest_cr_line          Jan-1985
last_credit_pull_d        Aug-2018
Name: 0, dtype: object


In [25]:
# As you see below, the home_ownership, verification_status, emp_length, term, and addr_state columns all contain multiple discrete values. 
# We should clean the emp_length column and treat it as a numerical one since the values have ordering

cols = ['home_ownership', 'verification_status', 'emp_length', 'term', 'addr_state']
for c in cols:
    print(loans[c].value_counts())

RENT        18471
MORTGAGE    17242
OWN          2837
OTHER          96
NONE            3
Name: home_ownership, dtype: int64
Not Verified       16468
Verified           12377
Source Verified     9804
Name: verification_status, dtype: int64
10    8897
0     4576
2     4389
3     4094
4     3435
5     3279
1     3240
6     2227
7     1771
8     1483
9     1258
Name: emp_length, dtype: int64
 36 months    28234
 60 months    10415
Name: term, dtype: int64
CA    6907
NY    3711
FL    2779
TX    2674
NJ    1825
IL    1487
PA    1481
VA    1378
GA    1358
MA    1313
OH    1190
MD    1034
AZ     832
WA     807
CO     769
NC     761
CT     734
MI     688
MO     661
MN     591
NV     482
SC     464
WI     445
OR     436
AL     433
LA     426
KY     323
OK     293
KS     260
UT     253
AR     235
DC     212
RI     197
NM     184
HI     169
WV     168
NH     162
DE     110
AK      79
WY      79
MT      79
SD      62
VT      54
MS      19
TN      17
IN       9
ID       6
IA       5
NE       5
ME  

In [26]:
print(loans["purpose"].value_counts())
print(loans["title"].value_counts())

debt_consolidation    18262
credit_card            5004
other                  3824
home_improvement       2884
major_purchase         2109
small_business         1783
car                    1497
wedding                 934
medical                 668
moving                  557
house                   369
vacation                351
educational             312
renewable_energy         95
Name: purpose, dtype: int64
Debt Consolidation                                    2149
Debt Consolidation Loan                               1695
Personal Loan                                          643
Consolidation                                          510
debt consolidation                                     489
Credit Card Consolidation                              349
Home Improvement                                       347
Debt consolidation                                     324
Small Business Loan                                    317
Credit Card Loan                                 

In [27]:
loans = loans.drop(['last_credit_pull_d', 'addr_state', 'title', 'earliest_cr_line'],axis=1)

In [28]:
# encode the home_ownership, verification_status, purpose, and term columns as dummy variables so we can use them in our model.

cat_columns = ["home_ownership", "verification_status", "purpose", "term"]
dummy_df = pd.get_dummies(loans[cat_columns])
loans_dm = pd.concat([loans, dummy_df], axis=1)
loans_dm = loans_dm.drop(cat_columns, axis=1)

In [29]:
loans_dm.head()

loan_amnt  int_rate  installment  emp_length  annual_inc  loan_status  \
0     5000.0    0.1065       162.87          10     24000.0            1   
1     2500.0    0.1527        59.83           0     30000.0            0   
2     2400.0    0.1596        84.33          10     12252.0            1   
3    10000.0    0.1349       339.31          10     49200.0            1   
4     3000.0    0.1269        67.79           1     80000.0            1   

     dti  delinq_2yrs  inq_last_6mths  open_acc       ...         \
0  27.65          0.0             1.0       3.0       ...          
1   1.00          0.0             5.0       3.0       ...          
2   8.72          0.0             2.0       2.0       ...          
3  20.00          0.0             1.0      10.0       ...          
4  17.94          0.0             0.0      15.0       ...          

   purpose_major_purchase  purpose_medical  purpose_moving  purpose_other  \
0                       0                0               0              0   
1                       0                0               0              0   
2                       0                0               0              0   
3                       0                0               0              1   
4                       0                0               0              1   

   purpose_renewable_energy  purpose_small_business  purpose_vacation  \
0                         0                       0                 0   
1                         0                       0                 0   
2                         0                       1                 0   
3                         0                       0                 0   
4                         0                       0                 0   

   purpose_wedding  term_ 36 months  term_ 60 months  
0                0                1                0  
1                0                0                1  
2                0                1                0  
3                0                1                0  
4                0                0                1  

[5 rows x 38 columns]

In [30]:
loans_dm.shape

(38649, 38)

In [31]:
import numpy

# This is how we build a model for prediction
# Predict that all loans will be paid off on time.

predictions = pd.Series(numpy.ones(loans_dm.shape[0]))

# False positives.
fp_filter = (predictions == 1) & (loans_dm["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_dm["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_dm["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_dm["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print('true positive rate : ',tpr)
print('false positive rate : ',fpr)

accuracy = (tp+tn)/(tp+tn+fp+fn)
print('accuracy :',accuracy)

true positive rate :  1.0
false positive rate :  1.0
accuracy : 0.8603832929072154


###  A good algorithm to apply to binary classification problems is logistic regression, for the following reasons:

** it's quick to train and we can iterate more quickly, **
** it's less prone to overfitting than more complex models like decision trees, **
** it's easy to interpret. **

In [32]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()

cols = loans_dm.columns
train_cols = cols.drop("loan_status")
features = loans_dm[train_cols]
target = loans_dm["loan_status"]

lr.fit(features, target)
predictions = lr.predict(features)

- Perform k-fold cross validation to get the accuracy of the model

In [33]:
from sklearn.model_selection import cross_val_predict
lr = LogisticRegression()
predictions = cross_val_predict(lr,features,target,cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_dm["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_dm["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_dm["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_dm["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp  / (tp + fn)
fpr = fp  / (fp + tn)


print('true positive rate : ',tpr)
print('false positive rate : ',fpr)

accuracy = (tp+tn)/(tp+tn+fp+fn)
print('accuracy :',accuracy)

true positive rate :  0.9993804021190248
false positive rate :  0.9988545246277205
accuracy : 0.8600101287416371


In [34]:
loans_dm.loan_status.value_counts()

1    33226
0     5423
Name: loan_status, dtype: int64

In [35]:
# There is imbalnce in the classes.
# The solution would be 
# 1. oversampling/undersampling 
# 2. penalizing misclassficationof the less prevalent class more than the other class
# with scikit-learn, we can easily apply the 2nd method 

In [36]:
from sklearn.model_selection import cross_val_predict

lr = LogisticRegression(class_weight="balanced")
predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)


# False positives.
fp_filter = (predictions == 1) & (loans_dm["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_dm["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_dm["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_dm["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print('true positive rate : ',tpr)
print('false positive rate : ',fpr)

accuracy = (tp+tn)/(tp+tn+fp+fn)
print('accuracy :',accuracy)

true positive rate :  0.6319588587007032
false positive rate :  0.6237113402061856
accuracy : 0.5962630274275662


In [37]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict
penalty = {
    0: 11,
    1: 1
}
# This above will impose a penalty of 6 for misclassifying a 0, and a penalty of 1 for misclassifying a 1.
# Since there are less 0.

lr = LogisticRegression(class_weight=penalty)
predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_dm["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans_dm["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_dm["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_dm["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print('true positive rate : ',tpr)
print('false positive rate : ',fpr)

accuracy = (tp+tn)/(tp+tn+fp+fn)
print('accuracy :',accuracy)

true positive rate :  0.062486446296353665
false positive rate :  0.06147384497899962
accuracy : 0.184796225711011


In [38]:
# I may use other algorithm RandomForestClassfier
# However, the result does not seem to imporive the model

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(class_weight="balanced", random_state=1)
predictions = cross_val_predict(rf, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans_dm["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.`
tp_filter = (predictions == 1) & (loans_dm["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans_dm["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans_dm["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print('true positive rate : ',tpr)
print('false positive rate : ',fpr)

accuracy = (tp+tn)/(tp+tn+fp+fn)
print('accuracy :',accuracy)

true positive rate :  0.9711886985346511
false positive rate :  0.9683085147002672
accuracy : 0.8400191912999441


** Classification model for Loan data **

In [39]:
x_data = loans.drop(['loan_status'],axis=1)
y_val = loans.loan_status

In [40]:
from sklearn.model_selection import train_test_split

In [41]:
X_train, X_test, y_train, y_test = train_test_split(x_data,y_val,test_size=0.3,random_state=101)

#### Scale the Feature Data

** Use sklearn preprocessing to create a MinMaxScaler for the feature data. Fit this scaler only to the training data. Then use it to transform X_test and X_train. Then use the scaled X_test and X_train along with pd.Dataframe to re-create two dataframes of scaled data.**

In [42]:
loans.head()

loan_amnt        term  int_rate  installment  emp_length home_ownership  \
0     5000.0   36 months    0.1065       162.87          10           RENT   
1     2500.0   60 months    0.1527        59.83           0           RENT   
2     2400.0   36 months    0.1596        84.33          10           RENT   
3    10000.0   36 months    0.1349       339.31          10           RENT   
4     3000.0   60 months    0.1269        67.79           1           RENT   

   annual_inc verification_status  loan_status         purpose    dti  \
0     24000.0            Verified            1     credit_card  27.65   
1     30000.0     Source Verified            0             car   1.00   
2     12252.0        Not Verified            1  small_business   8.72   
3     49200.0     Source Verified            1           other  20.00   
4     80000.0     Source Verified            1           other  17.94   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  revol_util  \
0          0.0             1.0       3.0      0.0    13648.0       0.837   
1          0.0             5.0       3.0      0.0     1687.0       0.094   
2          0.0             2.0       2.0      0.0     2956.0       0.985   
3          0.0             1.0      10.0      0.0     5598.0       0.210   
4          0.0             0.0      15.0      0.0    27783.0       0.539   

   total_acc  
0        9.0  
1        4.0  
2       10.0  
3       37.0  
4       38.0

In [43]:
from sklearn.preprocessing import MinMaxScaler

In [44]:
loans.iloc[1]

loan_amnt                         2500
term                         60 months
int_rate                        0.1527
installment                      59.83
emp_length                           0
home_ownership                    RENT
annual_inc                       30000
verification_status    Source Verified
loan_status                          0
purpose                            car
dti                                  1
delinq_2yrs                          0
inq_last_6mths                       5
open_acc                             3
pub_rec                              0
revol_bal                         1687
revol_util                       0.094
total_acc                            4
Name: 1, dtype: object

In [45]:
loans.dtypes

loan_amnt              float64
term                    object
int_rate               float64
installment            float64
emp_length               int64
home_ownership          object
annual_inc             float64
verification_status     object
loan_status              int64
purpose                 object
dti                    float64
delinq_2yrs            float64
inq_last_6mths         float64
open_acc               float64
pub_rec                float64
revol_bal              float64
revol_util             float64
total_acc              float64
dtype: object

In [46]:
import tensorflow as tf

D:\Users\KDH\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [47]:
loans.columns

Index(['loan_amnt', 'term', 'int_rate', 'installment', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'loan_status',
       'purpose', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
       'pub_rec', 'revol_bal', 'revol_util', 'total_acc'],
      dtype='object')

In [48]:
loan_amnt = tf.feature_column.numeric_column('loan_amnt')
term = tf.feature_column.categorical_column_with_vocabulary_list(key = 'term',vocabulary_list=[' 36 months', ' 60 months'], default_value=0 )
int_rate = tf.feature_column.numeric_column('int_rate')
installment = tf.feature_column.numeric_column('installment')
emp_length = tf.feature_column.numeric_column('emp_length')
home_ownership = tf.feature_column.categorical_column_with_hash_bucket("home_ownership",hash_bucket_size=5)
annual_inc = tf.feature_column.numeric_column('annual_inc')
verification_status = tf.feature_column.categorical_column_with_hash_bucket("verification_status",hash_bucket_size=3)
loan_status = tf.feature_column.numeric_column('loan_status')
purpose = tf.feature_column.categorical_column_with_hash_bucket('purpose',hash_bucket_size=14)
dti = tf.feature_column.numeric_column('dti')
delinq_2yrs = tf.feature_column.numeric_column('delinq_2yrs')
inq_last_6mths = tf.feature_column.numeric_column('inq_last_6mths')
open_acc = tf.feature_column.numeric_column('open_acc')
pub_rec = tf.feature_column.numeric_column('pub_rec')
revol_bal =tf.feature_column.numeric_column('revol_bal')
revol_util = tf.feature_column.numeric_column('revol_util')
total_acc =tf.feature_column.numeric_column('total_acc')

In [49]:
print(loans.term.unique())
print(loans.home_ownership.unique())
print(loans.verification_status.unique())
print(loans.purpose.unique())

[' 36 months' ' 60 months']
['RENT' 'OWN' 'MORTGAGE' 'OTHER' 'NONE']
['Verified' 'Source Verified' 'Not Verified']
['credit_card' 'car' 'small_business' 'other' 'wedding'
 'debt_consolidation' 'home_improvement' 'major_purchase' 'medical'
 'moving' 'vacation' 'house' 'renewable_energy' 'educational']


In [61]:
feat_cols = [loan_amnt,  int_rate, installment, emp_length,tf.feature_column.indicator_column(term),
       tf.feature_column.indicator_column(home_ownership), annual_inc, tf.feature_column.indicator_column(verification_status),tf.feature_column.indicator_column(purpose),dti, delinq_2yrs, inq_last_6mths, open_acc,pub_rec, total_acc]

### Create Input Function

** Batch_size is 100 with shuffle=True **

In [62]:
input_func=tf.estimator.inputs.pandas_input_fn(x=X_train,y=y_train,batch_size=100,num_epochs=None,shuffle=True)

### Create the model with tf.estimator

**Create a LinearClassifier.(Just in case for a DNNClassifier, I'll need to create embedded columns out of the cateogrical feature that use strings.)**

In [63]:
model = tf.estimator.LinearClassifier(feature_columns=feat_cols)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': 'C:\\Users\\KDH\\AppData\\Local\\Temp\\tmpguw0j6pc', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x0000028A6F3514E0>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


** Train the model on the data, for at least 5000 steps. **

In [64]:
model.train(input_fn=input_func,steps=5000)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1 into C:\Users\KDH\AppData\Local\Temp\tmpguw0j6pc\model.ckpt.
INFO:tensorflow:loss = 69.31472, step = 1
INFO:tensorflow:global_step/sec: 163.725
INFO:tensorflow:loss = 23078.533, step = 101 (0.614 sec)
INFO:tensorflow:global_step/sec: 203.165
INFO:tensorflow:loss = 2393.6362, step = 201 (0.490 sec)
INFO:tensorflow:global_step/sec: 206.943
INFO:tensorflow:loss = 22472.037, step = 301 (0.484 sec)
INFO:tensorflow:global_step/sec: 204.826
INFO:tensorflow:loss = 5579.2627, step = 401 (0.487 sec)
INFO:tensorflow:global_step/sec: 214.247
INFO:tensorflow:loss = 22020.887, step = 501 (0.467 sec)
INFO:tensorflow:global_step/sec: 200.923
INFO:tensorflow:loss = 3639.9956, step = 601 (0.498 sec)
INFO:tensorflow:global_step/sec

** Evaluation **

In [65]:
pred_fn = tf.estimator.inputs.pandas_input_fn(x=X_test,batch_size=len(X_test),shuffle=False)

In [66]:
predictions = list(model.predict(input_fn=pred_fn))

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\KDH\AppData\Local\Temp\tmpguw0j6pc\model.ckpt-5000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.


In [67]:
predictions[0]

{'class_ids': array([1], dtype=int64),
 'classes': array([b'1'], dtype=object),
 'logistic': array([1.], dtype=float32),
 'logits': array([217.27188], dtype=float32),
 'probabilities': array([0., 1.], dtype=float32)}

** Create a list of only the class_ids key values from the prediction list of dictionaries, these are the predictions you will use to compare against the real y_test values. **

In [68]:
final_preds = []
for pred in predictions:
    final_preds.append(pred['class_ids'][0])

In [69]:
final_preds[:15]

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

** Import classification_report from sklearn.metrics and then see if you can figure out how to use it to easily get a full report of your model's performance on the test data. **

In [70]:
from sklearn.metrics import classification_report

In [71]:
print(classification_report(y_test,final_preds))

             precision    recall  f1-score   support

          0       0.00      0.00      0.00      1623
          1       0.86      1.00      0.92      9972

avg / total       0.74      0.86      0.80     11595



- The precision is the ratio tp / (tp + fp) where tp is the number of true positives and fp the number of false positives. The precision is intuitively the ability of the classifier not to label as positive a sample that is negative.

-  The recall is the ratio tp / (tp + fn) where tp is the number of true positives and fn the number of false negatives. The recall is intuitively the ability of the classifier to find all the positive samples.

- The F-beta score can be interpreted as a weighted harmonic mean of the precision and recall, where an F-beta score reaches its best value at 1 and worst score at 0.

- The support is the number of occurrences of each class

-  Overall 80% seems ok but need to compare to other models. 
- and Given this, there's probably some room to improve. 
- I may change the batch_size, step size to draw the better result